File: src/languages/bavarian/bavarian.ts
/**
 * Bavarian (bar) phonemizer — Boarisch, Upper German (Austro-Bavarian), Latin script, canonical IPA,
 * espeak-independent. Bavarian has no codified state orthography, but the Bavarian Wikipedia / Wiktionary uses a
 * de-facto German-derived convention (⟨å⟩ for the dark [ɔ], ⟨ä ö ü⟩, ⟨ß⟩); that is what the wikipron referee
 * transcribes, so the g2p is well-defined against it. A greedy longest-match grapheme scan (diphthongs + digraphs,
 * length-desc) + code rules the table can't express:
 *   - the FALLING diphthongs ⟨ia ua oa⟩→[iɐ̯ uɐ̯ oɐ̯] (Biaschtl, Duafa, Boarisch) — Bavarian's hallmark — plus the closing
 *     ⟨ei⟩→[ɑɛ̯], ⟨au⟩→[ɑɔ̯], ⟨eu äu⟩→[ɔy̯] (offglides fold in the backbone);
 *   - the voiceless-LENIS stops ⟨b d g⟩→[b̥ d̥ ɡ̥] (under-ring; the Austro-Bavarian fortis/lenis contrast) vs fortis ⟨p t k⟩;
 *   - R-VOCALIZATION: a coda ⟨r⟩→[ɐ̯], the ending ⟨-er⟩→[ɐ] (Dreier→dräɐ, Wåssa-type);
 *   - the DORSAL FRICATIVE ⟨ch⟩→[ç] after a front vowel/consonant, [x] after a back vowel (the German ich/ach split);
 *   - initial ⟨st sp⟩→[ʃt ʃp], ⟨n⟩→[ŋ] before a velar, geminate collapse (no coda length contrast).
 * Vowel LENGTH and the fine dialect vowel qualities (ɑ~a~ɒ, e~e̞~ɛ) are folded/deferred; the referee is a NARROW
 * transcription with ~1.29 dialect variants/headword (credited any). See docs/investigations/bar_native_bringup_investigation.md.
 */
import type { Phonemizer } from "../../registry.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { hostWordRun, makeNativiser } from "../../core/hostWord.ts";
import { loadManifest } from "../../core/loadManifest.ts";
import { numberToWords } from "./numbers.ts";

interface BarDef {
    digraphs: Record<string, string>;
    graphemes: Record<string, string>;
    clausePunctuation: Record<string, string>;
}
const DEF = loadManifest<BarDef>(import.meta.url, "bavarian.jsonc");
const DIGRAPHS = DEF.digraphs;
const G = DEF.graphemes;
const CLAUSE_MARK = DEF.clausePunctuation;
// Multi-letter graphemes scanned longest-first so ⟨sch⟩ beats ⟨s⟩+⟨ch⟩, ⟨tsch⟩ beats ⟨t⟩+⟨sch⟩, etc.
const ORDER = Object.keys(DIGRAPHS).sort((a, b) => b.length - a.length);
const VOWEL_PH = new Set([..."aeiouyɑɔɛœøɒəɐ"]); // IPA vowel heads (for r-vocalization + ch front/back tests)
// A FRONT vowel head triggers the ⟨ch⟩→[ç] realization; anything else (a back vowel) → [x].
const FRONT_VOWEL = new Set([..."eiyɛœøɐ"]);

/** One scan token: the IPA phones for a grapheme + a flag marking a single ⟨r⟩ (may vocalize to [ɐ̯]/[ɐ] in a coda)
 *  and a single ⟨c⟩+⟨h⟩ ⟨ch⟩ (front/back realization, resolved in code). */
interface Tok { ph: string; rVar?: boolean; chVar?: boolean }

// The base characters of a phone, dropping combining diacritics (U+0300–036F incl. the ◌̯ offglide + the ◌̥ lenis
// under-ring + the ◌͡ tie) and the length mark ◌ː — so a diphthong (iɐ̯) / long vowel (aː) is still recognised as
// vowel-{initial,final} and a lenis [b̥] as its base [b].
function coreChars(ph: string): string[] {
    return [...ph.normalize("NFD")].filter((c) => {
        const x = c.codePointAt(0)!;
        return !(x >= 0x300 && x <= 0x36f) && x !== 0x2d0;
    });
}
function startsWithVowel(ph: string): boolean { const a = coreChars(ph); return a.length > 0 && VOWEL_PH.has(a[0]!); }
function endsWithVowel(ph: string): boolean { const a = coreChars(ph); return a.length > 0 && VOWEL_PH.has(a[a.length - 1]!); }
function endsWithFrontVowel(ph: string): boolean { const a = coreChars(ph); return a.length > 0 && FRONT_VOWEL.has(a[a.length - 1]!); }

/** Scan a lowercased Bavarian word into phone tokens (longest-match digraphs, then single graphemes). ⟨ch⟩ is kept
 *  as a marked token so its front/back realization can be resolved against the PRECEDING vowel in code. */
function scan(word: string): Tok[] {
    const w = word.normalize("NFC").toLowerCase();
    const toks: Tok[] = [];
    let i = 0;
    outer: while (i < w.length) {
        for (const key of ORDER) {
            if (w.startsWith(key, i)) {
                toks.push({ ph: DIGRAPHS[key]!, chVar: key === "ch" });
                i += key.length;
                continue outer;
            }
        }
        const c = w[i]!;
        const ph = G[c];
        if (ph !== undefined) toks.push({ ph, rVar: c === "r" });
        i += 1;
    }
    return toks;
}

/** The DORSAL FRICATIVE split: ⟨ch⟩ → [ç] (palatal) after a front vowel or word-initially/after a consonant, but
 *  [x] (velar) after a back vowel (the German ich/ach allophony; ⟨sch⟩ is [ʃ], handled by the digraph). */
function resolveCh(toks: Tok[]): void {
    for (let i = 0; i < toks.length; i++) {
        if (!toks[i]!.chVar) continue;
        const backContext = i > 0 && endsWithVowel(toks[i - 1]!.ph) && !endsWithFrontVowel(toks[i - 1]!.ph);
        toks[i]!.ph = backContext ? "x" : "ç";
    }
}

/** R-VOCALIZATION: a syllable-CODA ⟨r⟩ (no following vowel in the same or next segment) vocalizes to the offglide
 *  [ɐ̯]; the very common ending ⟨-er⟩ (a schwa+r after a consonant, at word end) → the syllabic vowel [ɐ]
 *  (Dreier→…ɐ, Wåssa-type). A prevocalic onset ⟨r⟩ (rot, Boarisch) stays the trill/tap [r]. */
function vocalizeR(toks: Tok[]): void {
    const last = toks.length - 1;
    const firstVowelIdx = toks.findIndex((t) => startsWithVowel(t.ph));
    for (let i = 0; i <= last; i++) {
        const t = toks[i]!;
        if (!t.rVar) continue;
        const nextVowel = i < last && startsWithVowel(toks[i + 1]!.ph);
        if (nextVowel) continue; // prevocalic onset → keep [r]
        // An UNSTRESSED ⟨-er⟩ (…e + coda r) collapses to a single [ɐ] — word-final (Dreier→…ɐ) or when the ⟨e⟩ is not
        // the first (stressed) vowel (Glumpert→…bɐt). A STRESSED ⟨er⟩ keeps the vowel + offglide (Bersch→b̥e̞ɐ̯ʃ).
        const prevIsE = i > 0 && toks[i - 1]!.ph === "e";
        if (prevIsE && (i === last || i - 1 !== firstVowelIdx)) { toks[i - 1]!.ph = "ɐ"; t.ph = ""; }
        else t.ph = "ɐ̯";
    }
}

/** Word-final ⟨gn⟩ [ɡ̥n] COALESCES to the velar nasal [ŋ] (Regn→reːŋ, Segn→seːŋ, Fliagn→fliɐ̯ŋ) — a regular Bavarian
 *  coda process (the preceding vowel lengthens, folded). The [ɡ̥] drops and the [n] becomes [ŋ]. */
function coalesceGN(toks: Tok[]): void {
    const last = toks.length - 1;
    if (last > 0 && toks[last]!.ph === "n" && toks[last - 1]!.ph === "ɡ̥") { toks[last - 1]!.ph = ""; toks[last]!.ph = "ŋ"; }
}

/** Initial ⟨st sp⟩ → [ʃt ʃp] (the German rule): a word-initial [s] before a coronal/labial stop becomes [ʃ]. The
 *  stop itself is the Bavarian LENIS [d̥]/[b̥] (⟨t p⟩ lenite, see the manifest), so the cluster surfaces as [ʃd̥]/[ʃb̥]. */
function initialSCluster(toks: Tok[]): void {
    if (toks.length >= 2 && toks[0]!.ph === "s" && (toks[1]!.ph === "d̥" || toks[1]!.ph === "b̥")) toks[0]!.ph = "ʃ";
}

/** POST-VOCALIC ⟨h⟩ is SILENT — a German vowel-length/hiatus marker, not the fricative [h] (Fruah→fruɐ̯, Fühn→fyn,
 *  Gfüh→ɡ̥fy). ⟨h⟩ keeps its [h] only as a syllable ONSET: word-initial or after a consonant, before a vowel. */
function silentH(toks: Tok[]): void {
    for (let i = 1; i < toks.length; i++) {
        if (toks[i]!.ph === "h" && endsWithVowel(toks[i - 1]!.ph)) toks[i]!.ph = "";
    }
}

/** ⟨k⟩ LENITES to [ɡ̥] everywhere EXCEPT a word-initial prevocalic onset — i.e. non-initially (coda/medial: Dackn→dɑɡ̥…,
 *  Unterflack→…flɑɡ̥, ⟨nk⟩ Dånkschee→…ŋɡ̥ʃ) and word-initially before a LIQUID (Klass→ɡ̥lɑs). This completes the
 *  Bavarian fortis/lenis neutralization for the velar (⟨t p⟩ lenite unconditionally in the manifest); only the
 *  word-initial prevocalic ⟨k⟩ keeps its fortis [k] (Kaas→kʰɑːs). */
function lenitK(toks: Tok[]): void {
    for (let i = 0; i < toks.length; i++) {
        if (toks[i]!.ph !== "k") continue;
        const initialBeforeLiquid = i === 0 && i + 1 < toks.length && (toks[i + 1]!.ph === "l" || toks[i + 1]!.ph === "r");
        if (i > 0 || initialBeforeLiquid) toks[i]!.ph = "ɡ̥";
    }
}

/** Word-final unstressed ⟨-a⟩ [ɑ] REDUCES to [ɐ] (Bana→b̥ɑːnɐ, Beda→b̥ɛːd̥ɐ, Oida→ɔed̥ɐ) — the Bavarian counterpart
 *  of the German final-schwa. Only a single final ⟨a⟩ after a consonant; the long ⟨aa⟩→[aː] and diphthong tails
 *  (already [ɐ̯]/[e]) are untouched. */
function reduceFinalA(toks: Tok[]): void {
    const last = toks.length - 1;
    if (last > 0 && toks[last]!.ph === "ɑ" && !endsWithVowel(toks[last - 1]!.ph)) toks[last]!.ph = "ɐ";
}

/** ⟨n⟩ → [ŋ] before a velar [k]/[ɡ̥] (the ⟨ng⟩ digraph is already [ŋ]; this catches ⟨nk⟩ Bånk→b̥ɔŋɡ̥, the [k]
 *  having lenited to [ɡ̥] by this point). */
function velarNasal(toks: Tok[]): void {
    for (let i = 0; i < toks.length - 1; i++) {
        if (toks[i]!.ph === "n" && (toks[i + 1]!.ph === "k" || toks[i + 1]!.ph === "ɡ̥")) toks[i]!.ph = "ŋ";
    }
}

/** Geminate collapse: a doubled consonant letter (⟨tt ll nn mm pp ff rr bb gg kk dd⟩) surfaces as a single phone
 *  (Bappn→bɑbm, Leffl→lefl) — Bavarian has no phonemic coda length contrast in this orthography. */
function degeminate(toks: Tok[]): void {
    for (let i = toks.length - 1; i > 0; i--) {
        const p = toks[i]!.ph;
        if (p !== "" && p === toks[i - 1]!.ph && !startsWithVowel(p)) toks.splice(i, 1);
    }
}

/** Phonemize a single Bavarian word to canonical IPA (segmental; fine vowel qualities + length folded/deferred). */
export function phonemizeWord(word: string): string {
    const toks = scan(word);
    resolveCh(toks);
    silentH(toks);
    initialSCluster(toks);
    // degeminate + coalesceGN run BEFORE lenitK: gemination must collapse while both stops are still identical, and
    // ⟨gn⟩-coalescence must see only a GENUINE ⟨g⟩→[ɡ̥] — not a ⟨k⟩ that lenitK turned into [ɡ̥] (else Dackn→d̥ɑŋ).
    degeminate(toks);
    coalesceGN(toks);
    lenitK(toks);
    velarNasal(toks);
    vocalizeR(toks);
    reduceFinalA(toks);
    return toks.map((t) => t.ph).join("");
}

// A word (Bavarian Latin letters incl. å ä ö ü é ß + accents) / number / punctuation token. ⟨ß⟩ is part of the
// de-facto Bavarian-Wikipedia orthography this engine reads (dreißg), so it must not split a word.
const TOKEN = new RegExp(`(${hostWordRun(["Latin"], "'-")})|(\\d+)|([.!?…,;:])`, "gu");

/**
 * This language's OWN inventory — the TOKEN word class as it stood before the widening above, lifted
 * verbatim, so nothing about the orthography is invented here. A token this REJECTS carries a letter the
 * language does not use, i.e. a foreign name. See core/hostWord.ts: this is the INVENTORY question, and it
 * is no longer also deciding where the script boundary falls (#657).
 */
const NATIVE_CLASS = "[a-zåäöüéèáàâßA-ZÅÄÖÜÉÈÁÀÂ'-]";
const nat = makeNativiser(NATIVE_CLASS, "u");

class BavarianPhonemizer implements Phonemizer {
    text(input: string): string {
        return assembleClauses(input, TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(nat(m[1])));
            // Numbers: the units-first compositor (numbers.ts) → each word back through the same g2p.
            else if (m[2]) for (const wd of numberToWords(Number(m[2])).split(" ")) sink.emit(phonemizeWord(wd));
            else if (m[3]) {
                const mk = CLAUSE_MARK[m[3]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Bavarian phonemizer (grapheme g2p + falling diphthongs + German-style rules; length/quality folded). */
export function createBavarian(): Phonemizer {
    return new BavarianPhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Bavarian/Bavarian.cs
/**
 * Bavarian (bar) phonemizer — Boarisch, Upper German (Austro-Bavarian), Latin script, canonical IPA. A
 * greedy longest-match grapheme scan plus the rules the table cannot express: the falling diphthongs
 * ⟨ia ua oa⟩, the voiceless-lenis stops ⟨b d g⟩→[b̥ d̥ ɡ̥], r-vocalization, the ich/ach dorsal split,
 * initial ⟨st sp⟩, ⟨gn⟩ coalescence, geminate collapse.
 *
 * Ported from src/languages/bavarian/bavarian.ts — see that file for the referee, the orthographic
 * convention it reads, and what is folded.
 */
using System.Text;
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Bavarian;

public static class BavarianPhonemizer
{
    private static BarDef DEF => Manifest.MANIFEST;
    private static IReadOnlyDictionary<string, string> DIGRAPHS => DEF.Digraphs;
    private static IReadOnlyDictionary<string, string> G => DEF.Graphemes;
    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => DEF.ClausePunctuation;

    // Longest-first, so ⟨sch⟩ beats ⟨s⟩+⟨ch⟩ and ⟨tsch⟩ beats ⟨t⟩+⟨sch⟩. JS's Array.prototype.sort is stable
    // and OrderByDescending is too; every remaining key is exactly two distinct characters, so at any one scan
    // position at most one of them can match and their relative order is immaterial.
    private static readonly string[] ORDER = DIGRAPHS.Keys.OrderByDescending(k => k.Length).ToArray();

    private static IReadOnlySet<string> VOWEL_PH => Ipa.IPA_VOWEL;

    /** A front vowel head triggers ⟨ch⟩→[ç]; anything else (a back vowel) → [x]. */
    private static readonly HashSet<string> FRONT_VOWEL = new(Js.CodePoints("eiyɛœøɐ"));

    /** One scan token: a grapheme's phones plus the two flags whose realization is resolved in code. */
    private sealed class Tok
    {
        public string Ph = "";
        public bool RVar;
        public bool ChVar;
    }

    /** A phone's base characters, dropping U+0300–036F and the length mark ◌ː, so a diphthong or long vowel
     *  is still recognised as vowel-{initial,final} and a lenis [b̥] as its base [b]. */
    private static List<string> CoreChars(string ph) =>
        Js.CodePoints(ph.Normalize(NormalizationForm.FormD))
            .Where(c =>
            {
                var x = Js.CodePointAt0(c);
                return !(x >= 0x300 && x <= 0x36f) && x != 0x2d0;
            })
            .ToList();

    private static bool StartsWithVowel(string ph)
    {
        var a = CoreChars(ph);
        return a.Count > 0 && VOWEL_PH.Contains(a[0]);
    }

    private static bool EndsWithVowel(string ph)
    {
        var a = CoreChars(ph);
        return a.Count > 0 && VOWEL_PH.Contains(a[^1]);
    }

    private static bool EndsWithFrontVowel(string ph)
    {
        var a = CoreChars(ph);
        return a.Count > 0 && FRONT_VOWEL.Contains(a[^1]);
    }

    /** Scan a lowercased word into phone tokens: longest-match digraphs, then single graphemes. */
    private static List<Tok> Scan(string word)
    {
        var w = Js.ToLowerCase(Js.Normalize(word, NormalizationForm.FormC));
        var toks = new List<Tok>();
        var i = 0;
        while (i < w.Length)
        {
            var matched = false;
            foreach (var key in ORDER)
            {
                // JS `w.startsWith(key, i)` is a UTF-16 code-unit comparison, which ordinal is.
                if (i + key.Length > w.Length || string.CompareOrdinal(w, i, key, 0, key.Length) != 0) continue;
                toks.Add(new Tok { Ph = DIGRAPHS[key], ChVar = key == "ch" });
                i += key.Length;
                matched = true;
                break;
            }
            if (matched) continue;
            var c = w[i].ToString();
            // ⚠ A letter with no rule here still denotes a sound; dropping it deletes what the writer typed.
            // Consulted AFTER every digraph and single-letter rule, so it cannot override this language.
            var ph = G.TryGetValue(c, out var g) ? g : LatinPhones.LatinPhone(c, new PhoneOpts { Initial = i == 0 });
            if (ph is not null) toks.Add(new Tok { Ph = ph, RVar = c == "r" });
            i += 1;
        }
        return toks;
    }

    /** The dorsal fricative split — the German ich/ach allophony. */
    private static void ResolveCh(List<Tok> toks)
    {
        for (var i = 0; i < toks.Count; i++)
        {
            if (!toks[i].ChVar) continue;
            var backContext = i > 0 && EndsWithVowel(toks[i - 1].Ph) && !EndsWithFrontVowel(toks[i - 1].Ph);
            toks[i].Ph = backContext ? "x" : "ç";
        }
    }

    /** A coda ⟨r⟩ vocalizes to [ɐ̯] and the ending ⟨-er⟩ to [ɐ]; a prevocalic onset ⟨r⟩ stays [r]. */
    private static void VocalizeR(List<Tok> toks)
    {
        var last = toks.Count - 1;
        var firstVowelIdx = toks.FindIndex(t => StartsWithVowel(t.Ph));
        for (var i = 0; i <= last; i++)
        {
            var t = toks[i];
            if (!t.RVar) continue;
            var nextVowel = i < last && StartsWithVowel(toks[i + 1].Ph);
            if (nextVowel) continue; // prevocalic onset → keep [r]
            // An unstressed ⟨-er⟩ collapses to one [ɐ]; a stressed one keeps vowel + offglide (Bersch).
            var prevIsE = i > 0 && toks[i - 1].Ph == "e";
            if (prevIsE && (i == last || i - 1 != firstVowelIdx)) { toks[i - 1].Ph = "ɐ"; t.Ph = ""; }
            else t.Ph = "ɐ̯";
        }
    }

    /** Word-final ⟨gn⟩ coalesces to [ŋ] (Regn→reŋ). */
    private static void CoalesceGN(List<Tok> toks)
    {
        var last = toks.Count - 1;
        if (last > 0 && toks[last].Ph == "n" && toks[last - 1].Ph == "ɡ̥")
        {
            toks[last - 1].Ph = "";
            toks[last].Ph = "ŋ";
        }
    }

    /** Initial ⟨st sp⟩ → [ʃd̥ ʃb̥] — the German rule over Bavarian's already-lenited stops. */
    private static void InitialSCluster(List<Tok> toks)
    {
        if (toks.Count >= 2 && toks[0].Ph == "s" && (toks[1].Ph == "d̥" || toks[1].Ph == "b̥")) toks[0].Ph = "ʃ";
    }

    /** Post-vocalic ⟨h⟩ is a silent length marker; it keeps its [h] only as a syllable onset. */
    private static void SilentH(List<Tok> toks)
    {
        for (var i = 1; i < toks.Count; i++)
            if (toks[i].Ph == "h" && EndsWithVowel(toks[i - 1].Ph)) toks[i].Ph = "";
    }

    /** ⟨k⟩ lenites to [ɡ̥] everywhere but a word-initial PREVOCALIC onset (Klass→ɡ̥lɑs, Kaas→kaːs). */
    private static void LenitK(List<Tok> toks)
    {
        for (var i = 0; i < toks.Count; i++)
        {
            if (toks[i].Ph != "k") continue;
            var initialBeforeLiquid =
                i == 0 && i + 1 < toks.Count && (toks[i + 1].Ph == "l" || toks[i + 1].Ph == "r");
            if (i > 0 || initialBeforeLiquid) toks[i].Ph = "ɡ̥";
        }
    }

    /** A single word-final ⟨-a⟩ after a consonant reduces to [ɐ] — the Bavarian final-schwa. */
    private static void ReduceFinalA(List<Tok> toks)
    {
        var last = toks.Count - 1;
        if (last > 0 && toks[last].Ph == "ɑ" && !EndsWithVowel(toks[last - 1].Ph)) toks[last].Ph = "ɐ";
    }

    /** ⟨n⟩ → [ŋ] before a velar — this catches ⟨nk⟩ (Bånk→b̥ɔŋɡ̥); ⟨ng⟩ is already [ŋ]. */
    private static void VelarNasal(List<Tok> toks)
    {
        for (var i = 0; i < toks.Count - 1; i++)
            if (toks[i].Ph == "n" && (toks[i + 1].Ph == "k" || toks[i + 1].Ph == "ɡ̥")) toks[i].Ph = "ŋ";
    }

    /** A doubled consonant letter surfaces as a single phone — no coda length contrast in this orthography. */
    private static void Degeminate(List<Tok> toks)
    {
        for (var i = toks.Count - 1; i > 0; i--)
        {
            var p = toks[i].Ph;
            if (p != "" && p == toks[i - 1].Ph && !StartsWithVowel(p)) toks.RemoveAt(i);
        }
    }

    /** Phonemize a single Bavarian word to canonical IPA. */
    public static string PhonemizeWord(string word)
    {
        var toks = Scan(word);
        ResolveCh(toks);
        SilentH(toks);
        InitialSCluster(toks);
        // ⚠ Degeminate and CoalesceGN run BEFORE LenitK: gemination must collapse while both stops are still
        // identical, and ⟨gn⟩-coalescence must see only a GENUINE ⟨g⟩→[ɡ̥] and not a ⟨k⟩ LenitK turned into
        // one (else Dackn→d̥ɑŋ).
        Degeminate(toks);
        CoalesceGN(toks);
        LenitK(toks);
        VelarNasal(toks);
        VocalizeR(toks);
        ReduceFinalA(toks);
        return string.Concat(toks.Select(t => t.Ph));
    }

    // A word / number / punctuation token.
    private static readonly JsRe TOKEN = JsRegex.Compile(
        $"({HostWord.HostWordRun(new[] { "Latin" }, "'-")})|(\\d+)|([.!?…,;:])", "gu");

    /** This language's OWN inventory — a different question from TOKEN's script boundary above. */
    private const string NATIVE_CLASS = "[a-zåäöüéèáàâßA-ZÅÄÖÜÉÈÁÀÂ'-]";
    private static readonly Func<string, string> Nat = HostWord.MakeNativiser(NATIVE_CLASS, "u");

    private sealed class Engine : ILanguage
    {
        public string Text(string input)
        {
            // Normalization first, so everything reaching TOKEN is something this g2p can speak.
            return Clauses.AssembleClauses(Normalize.NormalizeBavarian(input), TOKEN, (m, sink) =>
            {
                if (m.Groups[1].Success && m.Groups[1].Value.Length > 0)
                    sink.Emit(PhonemizeWord(Nat(m.Groups[1].Value)));
                // ⚠ THE TOKEN STRING IS PASSED AS `raw` (#1080): the digit-at-a-time fallback cannot recover
                // the digits from the double it exists to bypass. A bare `\d+` arm, so the token IS the digits.
                else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
                {
                    var tok = m.Groups[2].Value;
                    foreach (var wd in Numbers.NumberToWords(Js.Number(tok), tok).Split(' '))
                        sink.Emit(PhonemizeWord(wd));
                }
                else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
                {
                    if (CLAUSE_MARK.TryGetValue(m.Groups[3].Value, out var mk)) sink.Pause(mk);
                }
            });
        }
    }

    /** Build the Bavarian phonemizer. */
    public static ILanguage CreateBavarian() => new Engine();

    internal static void RegisterSelf() => Registry.Register("bavarian", () => CreateBavarian());
}
